### PR TITLE
Adds powerup system, refactors ctf pickups and powerup mine subtypes into it

### PR DESCRIPTION
**Pull request by Fikou**
## About The Pull Request

adds powerup system, when you bump/cross a powerup you trigger it, you can make the powerup respawn every few seconds if you want to
converts the yellow/red orb mine subtypes into powerups, refactors them to status effects
converts the ctf ammo pickup into powerup, makes it work with every non snowflake gun
adds a heal powerup, can be set to give you some amount of hp or give a fullheal

## Why It's Good For The Game

more consistent stuff, good for mappers possibly

## Changelog
:cl:
refactor: Refactors CTF pickups and mine powerup subtypes into a new powerup system!
/:cl:

### CODE REVIEW

File: code/game/objects/effects/powerup.dm [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	name = "power-up"
+	icon = 'icons/effects/effects.dmi'
+	density = FALSE
+	anchored = TRUE
+	resistance_flags = INDESTRUCTIBLE
+	var/respawning = FALSE
+	var/respawn_time
+	var/lifetime
+	var/pickup_message
+	var/pickup_sound
```
Comment (Jared-Fogle) [inline code/game/objects/effects/powerup.dm:11]: doc comments

File: code/game/objects/effects/powerup.dm [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	. = ..()
+	trigger(AM)
+
+/obj/effect/powerup/Bump(atom/A)
+	trigger(A)
+
+/obj/effect/powerup/Bumped(atom/movable/AM)
+	trigger(AM)
+
+/obj/effect/powerup/proc/trigger(mob/living/M)
```
Comment (Jared-Fogle) [inline code/game/objects/effects/powerup.dm:29]: better variable name

File: code/game/objects/effects/powerup.dm [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+		if(energygun.cell)
+			var/obj/item/stock_parts/cell/battery = energygun.cell
+			battery.charge = battery.maxcharge
+			energygun.update_icon()
+	for(var/obj/item/gun/ballistic/realgun in gear_list)
+		if(realgun.mag_type)
+			qdel(realgun.magazine)
+			realgun.magazine = new realgun.mag_type(realgun)
+			realgun.chamber_round()
+			realgun.update_icon()
```
Comment (Jared-Fogle) [inline code/game/objects/effects/powerup.dm:102]: I told you this is probably better as its own signal.

File: code/game/objects/effects/powerup.dm (hunk 2 above)
Comment (Jared-Fogle) [reply code/game/objects/effects/powerup.dm:29]: here and everywhere else

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Review by Mothblocks — APPROVED** [review]

**Comment by RaveRadbury** [comment]
Now we're playing with power!

**Review by Twaticus — APPROVED** [review]

File: code/modules/projectiles/guns/ballistic.dm [hunk 4]
```diff
@@ -516,10 +516,13 @@ GLOBAL_LIST_INIT(gun_saw_types, typecacheof(list(
 			. = TRUE
 
 /obj/item/gun/ballistic/proc/instant_reload()
-	if(!mag_type)
-		return
-	qdel(magazine)
-	magazine = new mag_type(src)
+	if(!magazine)
```
Comment (Jared-Fogle) [inline code/modules/projectiles/guns/ballistic.dm:519]: Invert this condition--you shouldn't be doing `if...else` on negative values.

**Comment by timothymtorres** [comment]
This broke ammo appearing in players pockets when they stepped on a reload powerup in #65571

```
/obj/effect/ctf/ammo/proc/reload(mob/living/M)
	if(!ishuman(M))
		return
	for(var/obj/machinery/capture_the_flag/CTF in GLOB.machines)
		if(M in CTF.spawned_mobs)
			var/outfit = CTF.spawned_mobs[M]
			var/datum/outfit/O = new outfit
			for(var/obj/item/gun/G in M)
				qdel(G)
			O.equip(M)
			to_chat(M, "<span class='notice'>Ammunition reloaded!</span>")
			playsound(get_turf(M), 'sound/weapons/gun/shotgun/rack.ogg', 50, TRUE, -1)
			qdel(src)
			break
```

**Pull request merged**